compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -923,10 +927,12 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
                         return Err(TypeError::IntrinsicCast);
                     }
 
-                    // Safe `#[target_feature]` functions are not assignable to safe fn pointers (RFC 2396).
+                    // Safe `#[target_feature]` functions are not assignable to safe fn pointers (RFC 2396),
+                    // report a better error than a safety mismatch.
+                    // FIXME(target_feature): do this inside `coerce_from_safe_fn`
 
-                    if b_hdr.safety == hir::Safety::Safe
-                        && !self.tcx.codegen_fn_attrs(def_id).target_features.is_empty()
+                    if b_hdr.safety.is_safe()
+                        && matches!(a_sig.safety(), hir::Safety::Unsafe { target_feature: true })
```

Comment:
As mentioned in https://github.com/rust-lang/rust/pull/134090#issuecomment-2539422624, we might want to make this actually safe if the function this happens in has the required features.
We probably shouldn't change it now, but we might want to modify the comment/FIXME accordingly

Reply:
If the features are also set via target information that can't differ between crates linked together?

Reply:
Not sure if doing so is supposed to be ok...

Anyway, I guess the point of the comment is that if you can call a function, then you also ought to be able to take a safe fn pointer to it.

Reply:
But safe pointers can be passed around to code where it's not safe to use them.

With the change I made (plus encoding the actually used target features) we can in theory start creating safe target feature pointers by keeping the target feature information on the function pointer type.

Reply:
target_feature_11 enables this:
```
#[target_feature(enable = "avx2")]
fn bar() -> fn() {
    || avx2()
}
```
which is not particularly different from allowing coercion of `avx2` to a fn pointer. If you have concerns about this, the stabilization PR is probably the better place to discuss it (the stabilization report mentions why it's OK to do this)

Reply:
Oh right, yea that makes sense. You already had to call an unsafe function to get here, and it's not a footgun, because it's global informatiom, not local